Fix #4005: Move ItemLoader to a new library

ejulio opened the pull request:
Fix #4005

Installed the new library [`itemloaders`](https://github.com/scrapy/itemloaders) from github for now.
Remove tests that are already in `itemloaders`. Kept some that seem to be a bit specific to scrapy. Not completely on this removal. Thoughts?

Made scrapy's `ItemLoader` an extension of the base class from the library an added support for `Item`

Removed processors and imported them from `itemloaders`

Removed `common.py` with `wrap_loader_context` which moved to the new library

Before merging:

- [x] Define the new library name 
    - `itemloaders`
    - `data_loaders`: move away from scrapy's idea of items
    - `xhtml_loaders`: loads data from any XHTML text/parsel selector
- [x] Update scrapy docs mentioning the library

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,219 +3,35 @@
     default_selector_class = Selector
 
-    def __init__(self, item=None, selector=None, response=None, parent=None, **context):
+    def __init__(self, response=None, item=None, selector=None, parent=None, **context):
```
Can we keep the previous parameter order, for backward compatibility?

victor-torres replied:
Any reason for changing arguments order here? It might break code that doesn't make use of named args.

Gallaecio commented on scrapy/loader/common.py:
```diff
@@ -1,14 +0,0 @@
-"""Common functions used in Item Loaders code"""
-
-from functools import partial
-from scrapy.utils.python import get_func_args
-
-
-def wrap_loader_context(function, context):
-    """Wrap functions that receive loader_context to contain the context
-    "pre-loaded" and expose a interface that receives only one argument
-    """
-    if 'loader_context' in get_func_args(function):
-        return partial(function, loader_context=context)
-    else:
-        return function
```
To maximize backward compatibility, we might want to deprecate this module instead of removing it. Maybe we can import the function from the new library, though.

ejulio replied:
I guess keeping the import is better than adding a deprecation, thoughts?
Also, I removed it as it doesn't appear in the docs https://docs.scrapy.org/en/latest/topics/loaders.html, so it shouldn't be used outside `ItemLoader`, or am I missing something here?

kmike replied:
Yes, an import + deprecation warning. 

We're usually trying to be quite careful about deprecations, and even parts which were not publicly documented are going through deprecation process. 

Here you can also make an argument that the module is not private (not _common), and the function  is not private (not _wrap_loader_context).

Gallaecio commented on scrapy/loader/processors.py:
```diff
@@ -1,104 +1 @@
-
-    def __call__(self, values):
-        return self.separator.join(values)
+from itemloaders.processors import *
```
We need to issue a deprecation warning, to try and get people importing from `itemloaders` directly in the long term.

wRAR replied:
We don't do a similar thing for `parsel`, do we?

Gallaecio replied:
Because we have custom `Selector` subclasses in that case.

kmike replied:
we do it for w3lib imports, e.g. https://github.com/scrapy/scrapy/blob/master/scrapy/utils/http.py (some other cases like that have been removed)

kmike replied:
or https://github.com/scrapy/scrapy/blob/master/scrapy/utils/markup.py

ejulio replied:
Here we have both options.
Consider a deprecation and help people move to `itemloaders`, or just import it as part of `scrapy`.
Which one is preferred?

kmike replied:
I think we should issue a deprcation warning and ask people to use itemloaders.

victor-torres commented on scrapy/loader/processors.py:
```diff
@@ -1,104 +1 @@
-
-    def __call__(self, values):
-        return self.separator.join(values)
+from itemloaders.processors import *
```
Maybe it's better to import each processor individually. This way we keep a text reference to them instead of a dynamic "import all".

ejulio replied:
I guess this is fine, though, for any new processor, we'll need to update this piece of code.
Unless, we import them with some deprecation warning as mentioned above.

Gallaecio replied:
Actually, since the goal is for people to import directly from `itemloaders` (which is why this module should yield a warning when imported), even if `itemloaders` adds new processors we would not want them imported here.

kmike commented on tests/test_loader.py:
```diff
@@ -630,367 +227,6 @@ def test_avoid_reprocessing_without_initial_values_list(self):
-        self.assertEqual(l.get_output_value('name'), [u'Marta'])
-
-    def test_init_method_with_response(self):
-        l = TestItemLoader(response=self.response)
```
I think we should be keeping tests like that. This removal probably needs a more detailed review: everything which is related to response, or e.g. scrapy.Item, should be tested, because these are features which Scrapy's wrapper provides, not features which itemloaders library provides.

I also wonder if many (most) other tests needs to be preserved: e.g. I can see how we could have implemented a scrapy ItemLoader subclass in a way which breaks itemloaders.ItemLoader's `selector` argument. Though we haven't done this for scrapy.Selector, so no strong opinion here.

ejulio replied:
Makes sense to me.
The ones I removed were related to a usage of `response` in which I transformed to `selector` in `itemloaders`.
But I'm in favor of keeping the ones referencing `response`..

kmike commented on scrapy/loader/processors.py:
```diff
@@ -1,104 +1 @@
-        return values
-
-
-class SelectJmes:
```
If I'm not mistaken, this is the only place where jmespath is used in Scrapy; I wonder if its mentions should be removed (from docs, from CI?).

ejulio replied:
Removed the dependency.
I'll remove from the docs once I update it referencing the librayr

kmike commented:
As for the package name, my vote is for `itemloaders`.

ejulio commented:
@kmike @Gallaecio @wRAR @victor-torres 
Thanks for the comments.
JUst left some thoughts on the deprecation and imports from the new library.
Once we get it sorted out, I'll update it.

ejulio commented:
@victor-torres @kmike @Gallaecio @wRAR 
I updated the changes to log warnings whenever something that was ported is used.
Not sure if that's the best approach, so fell free to raise any concerns about it.

Also, updated the documentation removing most of the content.
The idea is to include a link to `itemloaders` library docs once it is up and running.
Regarding the docs, not sure if we should keep some sort of TOC pointing to stuff that was moved from scrapy..

kmike commented on docs/topics/loaders.rst:
```diff
@@ -179,595 +114,3 @@ output processors are declared using the ``_out`` suffix. And you can also
-ItemLoader objects
-==================
-
-.. class:: ItemLoader([item, selector, response], \**kwargs)
```
Is this signature documented anywhere now?

kmike commented on docs/topics/loaders.rst:
```diff
@@ -179,595 +114,3 @@ output processors are declared using the ``_out`` suffix. And you can also
-    given, one is instantiated automatically using the class in
-    :attr:`default_item_class`.
-
-    When instantiated with a ``selector`` or a ``response`` parameters
```
I'm not sure this is documented anywhere, after the removal

kmike commented on docs/topics/loaders.rst:
```diff
@@ -179,595 +114,3 @@ output processors are declared using the ``_out`` suffix. And you can also
-        (resp. :meth:`replace_css`) method.
-    :type selector: :class:`~scrapy.selector.Selector` object
-
-    :param response: The response used to construct the selector using the
```
this is not documented now

kmike commented on docs/topics/loaders.rst:
```diff
@@ -161,10 +92,14 @@ Item Loaders are declared like Items, by using a class definition syntax. Here
 
+        default_item_class = ProductItem
+
         default_output_processor = TakeFirst()
```
I think this example is not clear after processors docs are removed

ejulio replied:
This one is a bit tricky.
I included a "link" to `itemloaders` doc after the example.
Not sure if it's the best approach

kmike commented on scrapy/loader/processors.py:
```diff
@@ -4,101 +4,61 @@
 
 
-class MapCompose:
+class MapCompose(processors.MapCompose):
```
I wonder if you can use scrapy.utils.deprecate.create_deprecated_class here

kmike commented on docs/topics/loaders.rst:
```diff
@@ -179,595 +114,3 @@ output processors are declared using the ``_out`` suffix. And you can also
-
-    :class:`ItemLoader` instances have the following methods:
-
-    .. method:: get_value(value, \*processors, \**kwargs)
```
A proposal: to avoid copy-paste, but still have docs for these methods in Scrapy itself, make them docstrings in itemloaders repo, and then use autodocs. If I'm not mistaken, this is what we've done with scrapy.Selector, when it was moved to parsel.

elacuesta commented on docs/topics/loaders.rst:
```diff
@@ -68,8 +68,11 @@ previously extracted and collected with the :meth:`~ItemLoader.add_xpath`,
 
-Item Loaders are declared like Items, by using a class definition syntax. Here
-is an example::
+Besides the usage of Item Laoders as instances, like the example above,
```
```suggestion
Besides the usage of Item Loaders as instances, like the example above,
```

elacuesta commented on docs/topics/loaders.rst:
```diff
@@ -77,91 +64,18 @@ called which actually returns the item populated with the data
 
-Item Loaders are declared like Items, by using a class definition syntax. Here
-is an example::
+Besides the usage of Item Laoders as instances, like the example above,
```
```suggestion
Besides the usage of Item Loaders as instances, like the example above,
```

Gallaecio commented on docs/conf.py:
```diff
@@ -288,6 +288,7 @@
     'tox': ('https://tox.readthedocs.io/en/latest', None),
     'twisted': ('https://twistedmatrix.com/documents/current', None),
     'twistedapi': ('https://twistedmatrix.com/documents/current/api', None),
+    'itemloaders': ('https://itemloaders.readthedocs.io/en/latest/', None),
```
:lipstick: Let’s keep it alphabetical :nerd_face:

Gallaecio commented on scrapy/loader/common.py:
```diff
@@ -1,14 +1,19 @@
 """Common functions used in Item Loaders code"""
 
 from functools import partial
+import warnings
```
:lipstick: 

```suggestion
import warnings
from functools import partial
```

Gallaecio commented on scrapy/loader/common.py:
```diff
@@ -1,14 +1,19 @@
+
+from itemloaders import common
+
 from scrapy.utils.python import get_func_args
+from scrapy.utils.deprecate import ScrapyDeprecationWarning
```
:lipstick: 

```suggestion
from scrapy.utils.deprecate import ScrapyDeprecationWarning
from scrapy.utils.python import get_func_args
```

Gallaecio commented on scrapy/loader/common.py:
```diff
@@ -1,14 +1,19 @@
 
 
 def wrap_loader_context(function, context):
-    """Wrap functions that receive loader_context to contain the context
-    "pre-loaded" and expose a interface that receives only one argument
-    """
```
It should be OK to leave the docstring until we remove the function.

Gallaecio commented on tests/test_loader_deprecated.py:
```diff
@@ -0,0 +1,695 @@
+"""
```
Given `ItemLoader` is no longer deprecated, it may make sense to restore the original file name.

ejulio replied:
Maybe a better name, these are the tests that were moved to the new library.
The idea is to remove them once the deprecated stuff (processors, ...) is removed.
`test_loader_moved_itemloaders.py`?

Gallaecio replied:
Or maybe we can leave the `test_loader_deprecated` name but add a `test_loader` with the tests covering Scrapy-specific features that will remain after those deprecations.

ejulio replied:
@Gallaecio , but there's already a file named `test_loader.py`, which is the past one, but I removed the tests that were covered by `itemloaders` (moved them to `test_loader_deprecated`)

Gallaecio replied:
:facepalm: 

My bad, it did not load by default in the patch viewer and because it only had deleted lines in the count of added and deleted lines I though the whole file had been deleted.

kmike commented on docs/topics/loaders.rst:
```diff
@@ -174,600 +88,31 @@ is an example::
-
-1. Item Loader field-specific attributes: ``field_in`` and ``field_out`` (most
-   precedence)
-2. Field metadata (``input_processor`` and ``output_processor`` key)
```
I could be wrong, but it seems this information is absent in new docs.

ejulio replied:
It was moved to `itemloaders`, but the `Item` part was missing.

kmike commented on scrapy/loader/common.py:
```diff
@@ -1,14 +1,22 @@
+    warnings.warn(
+        "scrapy.loader.common.wrap_loader_context has moved to a new library."
+        "Please update your reference to itemloaders.common.wrap_loader_context",
+        ScrapyDeprecationWarning
```
should it use stacklevel=2?

ejulio replied:
TIL :+1:

kmike commented on docs/topics/loaders.rst:
```diff
@@ -174,600 +88,31 @@ is an example::
-the Item Loader definition, and it's very common to declare input processors
-this way. However, there is one more place where you can specify the input and
-output processors to use: in the :ref:`Item Field <topics-items-fields>`
-metadata. Here is an example::
```
Is there an example of using Item meta for declaring of input and output processors in the new docs?

kmike commented on docs/topics/loaders.rst:
```diff
@@ -174,600 +88,31 @@ is an example::
-        It's either the selector given in the ``__init__`` method or one created from
-        the response given in the ``__init__`` method using the
-        :attr:`default_selector_class`. This attribute is meant to be
-        read-only.
```
Are the docs for loader.selector attribute available? They differ in itemloaders and scrapy.loader, because the selector can be created from response in scrapy's wrapper.

Gallaecio commented on docs/topics/loaders.rst:
```diff
@@ -21,26 +21,13 @@ Item Loaders are designed to provide a flexible, efficient and easy mechanism
 for extending and overriding different field parsing rules, either by spider,
 or by source format (HTML, XML, etc) without becoming a nightmare to maintain.
 
+.. note:: Item Loaders used to be a part of ``scrapy``, but moved to a self-contained library.
+    This implementation is an extension of itemloaders_ library,
+    keeping support to :ref:`Items <topics-items>` and :ref:`Response <topics-request-response>`.
+
```
I think this should be written more in line with the counterpart note in `selectors.rst`, talking only about the present, not the past.

Gallaecio commented on docs/topics/loaders.rst:
```diff
@@ -623,151 +388,7 @@ needs.
-    data is often extracted using the
-    :meth:`~scrapy.selector.Selector.extract` method of :ref:`selectors
-    <topics-selectors>`, which returns a list of unicode strings.
+.. automodule:: itemloaders.processors
+    :members:
```
Since this is non-Scrapy API reference documentation, I would be OK with removing this section, and instead updating the paragraph above which mentions that `itemloaders` provides built-in processors, so that it links to the upstream documentation about `itemloaders` built-in processors.

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,219 +3,104 @@
+        read-only.
+    """
 
     default_item_class = Item
```
Is this useful? I ask because I imagine that, given `Item` has no fields defined, this always needs to be overriden with a different value.

ejulio replied:
This is here only for backwards compatibility, as this is how it was before.
I guess there should be no harm in removing it

Gallaecio replied:
Right, backward compatibility. Keeping it makes sense.

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,219 +3,104 @@
-        self._check_selector_method()
-        csss = arg_to_iter(csss)
-        return flatten(self.selector.css(css).getall() for css in csss)
+            return self.item.fields[field_name].get(key)
```
I guess we need to wait for #3881 and use `ItemAdapter` here. Maybe also in the `itemloaders` library itself?

kmike replied:
That's the question, what should we merge first :)

ejulio replied:
Both work for me.
Just need to keep in mind that, as we wait, new changes may come to the files that were moved, requiring to patch/merge changes in the new repo..
This can get a bit complicated, as we merged the changes there (so we may need to go back in history to merge and then apply the changes back again)

kmike commented on setup.py:
```diff
@@ -71,6 +71,7 @@ def has_environment_marker_platform_impl_support():
         'Twisted>=17.9.0',
         'cryptography>=2.0',
         'cssselect>=0.9.1',
+        'itemloaders>=1.0.0',
```
Could you please check that itemloader's setup.py uses the same minimal package requirements as Scrapy?

victor-torres replied:
What do you mean by "the same minimal package requirements as Scrapy"?

kmike replied:
If itemloaders require e.g. parsel 1.6, and scrapy requires 1.5, and now scrapy requires itemloaders, than we're implicitly bumping minimal supported parsel version in scrapy.

ejulio replied:
I'll update the requirements there..
Though, `jmespath` was optional in scrapy, but is a requirement in `itemloaders`
It shouldn't be a problem, but it changed.

ejulio replied:
https://github.com/scrapy/itemloaders/pull/10

kmike commented on scrapy/loader/__init__.py:
```diff
@@ -3,219 +3,104 @@
-        if '.' not in method.__qualname__:
-            return method.__func__
-    return method
+    Return a new Item Loader for populating the given Item. If no item is
```
It seems this docstring explains how `ItemLoader.__init__` works, not what is ItemLoader class about. Could it be moved to `__init__` method (is autodocs output reasonable?)?

Alternatively, it can be rewritten as a class docstring. This makes more sense to me, because e.g. attribute docs don't really belong to `__init__` docstring.

kmike commented on scrapy/loader/__init__.py:
```diff
@@ -3,219 +3,103 @@
+        context.update(response=response)
+        super().__init__(item=item, selector=selector, parent=parent, **context)
+
+    def get_default_input_processor_for_field(self, field_name):
```
hm, I suspect these extension points won't be necessary once itemloader gets itemadapter support.

ejulio replied:
That's true

Gallaecio replied:
Itemadapter support is in.

noviluni commented:
Adding this link here to generate a reference of this in that issue: https://github.com/scrapy/scrapy/issues/1777#issuecomment-184272838

Gallaecio commented:
Regarding the failed `typing` job, we are not enforcing type annotations at the moment, so feel free to just add the affected files to the exception list in `setup.cfg`.

ejulio commented:
Thanks for the heads-up @Gallaecio .
Was trying to understand the reasons for the failure.

Gallaecio commented on docs/topics/loaders.rst:
```diff
@@ -20,6 +20,10 @@ Item Loaders are designed to provide a flexible, efficient and easy mechanism
 or by source format (HTML, XML, etc) without becoming a nightmare to maintain.
 
+.. note:: Item Loaders are an extension of the itemloaders_ library that make it
+    easier to work with Scrapy by adding support for :ref:`Items <topics-items>`
```
:lipstick: 

```suggestion
    easier to work with Scrapy by adding support for :ref:`items <topics-items>`
```

Gallaecio commented on docs/topics/loaders.rst:
```diff
@@ -173,17 +177,17 @@ The other thing you need to keep in mind is that the values returned by input
 
-Last, but not least, Scrapy comes with some :ref:`commonly used processors
-<topics-loaders-available-processors>` built-in for convenience.
+Last, but not least, ``itemloaders`` comes with some :ref:`commonly used
+processors <processors>` built-in for convenience.
```
```suggestion
Last, but not least, itemloaders_ comes with some :ref:`commonly used
processors <itemloaders:built-in-processors>` built-in for convenience.
```

Note: ``` :doc:`itemloaders <itemloaders:index>` ``` could work as an alternative to `itemloaders_` that links to the home page of the `itemloaders` documentation.

Gallaecio commented on docs/topics/loaders.rst:
```diff
@@ -213,8 +217,8 @@ this way. However, there is one more place where you can specify the input and
 output processors to use: in the :ref:`Item Field <topics-items-fields>`
 metadata. Here is an example::
 
+    from itemloaders.processors import Join, MapCompose, TakeFirst
     import scrapy
```
:lipstick: 

```suggestion
    import scrapy
    from itemloaders.processors import Join, MapCompose, TakeFirst
```

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,217 +3,86 @@
-        if '.' not in method.__qualname__:
-            return method.__func__
-    return method
+    A user-fiendly abstraction to populate data to an :ref:`Item <topics-items>`
```
```suggestion
    A user-friendly abstraction to populate an :ref:`item <topics-items>` with data
```

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,217 +3,86 @@
+    :param item: The item instance to populate using subsequent calls to
+        :meth:`~ItemLoader.add_xpath`, :meth:`~ItemLoader.add_css`,
+        or :meth:`~ItemLoader.add_value`.
+    :type item: :class:`~scrapy.item.Item` object
```
I believe this syntax works as well:

```suggestion
    :type item: scrapy.item.Item
```

No strong opinion, though.

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,217 +3,86 @@
+    :type item: :class:`~scrapy.item.Item` object
+
+    :param selector: The selector to extract data from, when using the
+        :meth:`add_xpath` (resp. :meth:`add_css`) or :meth:`replace_xpath`
```
What does “resp.” mean here?

:lipstick: Maybe we could just list the 4 methods, as a plain list.

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,217 +3,86 @@
+    :type response: :class:`~scrapy.http.Response` object
+
+    If no item is given, one is instantiated automatically using the class in
+    :attr:`default_item_class`.
+    The item, selector, response and the remaining keyword arguments are
```
```suggestion
    :attr:`default_item_class`.

    The item, selector, response and remaining keyword arguments are
```

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,217 +3,86 @@
+    .. attribute:: item
+
+        The item object being parsed by this Item Loader.
+        This is mostly used as a property so when attempting to override this
```
```suggestion
        This is mostly used as a property so, when attempting to override this
```

Gallaecio commented on scrapy/loader/__init__.py:
```diff
@@ -3,217 +3,86 @@
+    .. attribute:: default_item_class
 
-class ItemLoader:
+        An Item class (or factory), used to instantiate items when not given in
```
If `dict` is a supported value, even if `attr.s` and `dataclass` are experimental, maybe we should say lowercase ‘item’ with a link to the topic on items.

kmike commented on docs/topics/loaders.rst:
```diff
@@ -20,6 +20,10 @@ Item Loaders are designed to provide a flexible, efficient and easy mechanism
 or by source format (HTML, XML, etc) without becoming a nightmare to maintain.
 
+.. note:: Item Loaders are an extension of the itemloaders_ library that make it
+    easier to work with Scrapy by adding support for :ref:`items <topics-items>`
```
Aren't items supported in itemloaders out of box, via itemadapter?

ejulio replied:
Yup, I'll leave only the reference to `response`

Gallaecio commented on tests/test_loader.py:
```diff
@@ -7,8 +6,7 @@
 from scrapy.loader import ItemLoader
-from scrapy.loader.processors import (Compose, Identity, Join,
-                                      MapCompose, SelectJmes, TakeFirst)
+from scrapy.loader.processors import Compose, Identity, MapCompose, TakeFirst
```
In this and any other file were we import processors, we should start importing them from the new library.

Gallaecio commented on tests/test_loader_deprecated.py:
```diff
@@ -0,0 +1,695 @@
+Once we remove the references from scrapy, we can remove these tests.
+"""
+
+from functools import partial
+import unittest
```
:lipstick: 

```suggestion
import unittest
from functools import partial
```

ejulio commented:
@Gallaecio , I've updated the PR with your ideas.
Just bear in mind in that, `extract_regex` was already in `parsel.utils`, so I just added a simple test in `test_loader_deprecated` so we don't need a new file only for this case.

ejulio commented:
I can't believe this was finally merged :tada: :stuck_out_tongue: 
Thanks for all the effort here @victor-torres @Gallaecio @kmike :clap:
